# Homework 3
### Rolando Santos
### https://github.com/rsantos2032/DSCI401/
### 2023-10-03

In [1]:
# Loading the Dataset
import pandas as pd
Batting = pd.read_csv("https://raw.githubusercontent.com/gjm112/DSCI401/main/data/Batting.csv")
Pitching = pd.read_csv("https://raw.githubusercontent.com/gjm112/DSCI401/main/data/Pitching.csv")
People = pd.read_csv("https://raw.githubusercontent.com/gjm112/DSCI401/main/data/People.csv")

### Question 1: Use the Batting, Pitching, and People tables in the Lahman package to answer the following questions:

#### a. Name every player in baseball history who has accumulated at least 300 home runs (HR) AND at least 300 stolen bases (SB). You can find the first and last name of the player in the People data frame. Join this to your result along with the total home runs and total bases stolen for each of these elite players.

In [27]:
Batting_ = Batting.groupby('playerID').agg({'SB':'sum','HR':'sum'}).reset_index()
Batting_ = Batting_.loc[(Batting_['HR'] >= 300) & (Batting_['SB'] >= 300)]
Batting_ = pd.merge(Batting_, People, left_on = "playerID", right_on = "playerID", how = "left")
Batting_ = Batting_.loc[:, ["playerID", "nameFirst", "nameLast", "SB", "HR"]]
Batting_

playerID nameFirst   nameLast     SB   HR
0  beltrca01    Carlos    Beltran  312.0  435
1  bondsba01     Barry      Bonds  514.0  762
2  bondsbo01     Bobby      Bonds  461.0  332
3  dawsoan01     Andre     Dawson  314.0  438
4  finlest01     Steve     Finley  320.0  304
5   mayswi01    Willie       Mays  338.0  660
6  rodrial01      Alex  Rodriguez  329.0  696
7  sandere02    Reggie    Sanders  304.0  305

#### b. Similarly, name every pitcher in baseball history who has accumulated at least 300 wins (W) and at least 3,000 strikeouts (SO).

In [33]:
Pitching_ = Pitching.groupby('playerID').agg({'W':'sum','SO':'sum'}).reset_index()
Pitching_ = Pitching_.loc[(Pitching_['W'] >= 300) & (Pitching_['SO'] >= 3000)]
Pitching_ = pd.merge(Pitching_, People, left_on = 'playerID', right_on = 'playerID', how = 'left')
Pitching_ = Pitching_.loc[:, ['playerID', 'nameFirst', 'nameLast', 'W', 'SO']]
Pitching_

playerID nameFirst nameLast    W    SO
0  carltst01     Steve  Carlton  329  4136
1  clemero02     Roger  Clemens  354  4672
2  johnsra05     Randy  Johnson  303  4875
3  johnswa01    Walter  Johnson  417  3509
4  maddugr01      Greg   Maddux  355  3371
5  niekrph01      Phil   Niekro  318  3342
6  perryga01   Gaylord    Perry  314  3534
7   ryanno01     Nolan     Ryan  324  5714
8  seaveto01       Tom   Seaver  311  3640
9  suttodo01       Don   Sutton  324  3574

#### c. Identify the name and year of every player who has hit at least 50 home runs in a single season. Which player had the lowest batting average in that season?

In [49]:
Batting_ = Batting.groupby(['playerID', 'yearID']).sum().reset_index()
Batting_['BA'] = Batting_['H']/Batting_['AB']
Batting_ = Batting_.loc[Batting_['HR'] >= 50]
Batting_ = pd.merge(Batting_, People, left_on = 'playerID', right_on = 'playerID', how = 'left')
Batting_ = Batting_.loc[:, ['playerID', 'yearID', 'HR', 'BA', 'nameFirst', 'nameLast']]
Batting_.sort_values(by = 'BA', ascending = True).head()

playerID  yearID  HR        BA nameFirst  nameLast
0   alonspe01    2019  53  0.259631      Pete    Alonso
2   bautijo02    2010  54  0.260105      Jose  Bautista
16  jonesan01    2005  51  0.262799    Andruw     Jones
22  marisro01    1961  61  0.269492     Roger     Maris
44  vaughgr01    1998  50  0.272251      Greg    Vaughn

From the dataset, it appears that Pete Alonso had the lowest batting average in the 2019 season.

### Question 2: Use the nycflights13 package and the flights and planes tables to answer the following questions:

In [94]:
planes = pd.read_csv("https://raw.githubusercontent.com/gjm112/DSCI401/main/data/planes.csv")
flights = pd.read_csv("https://raw.githubusercontent.com/gjm112/DSCI401/main/data/flights.csv")

#### a. What is the oldest plane (specified by the tailnum variable) that flew from New York City airports in 2013?

In [100]:
planes_ = planes.rename(columns={'year': 'created'})
planes_['age'] = 2013 - planes_['created']
planes_ = pd.merge(planes_, flights, left_on = 'tailnum', right_on = 'tailnum', how = 'left')
planes_ = planes_.loc[:, ['tailnum', 'created', 'age']]
planes_ = planes_.sort_values(by = 'age', ascending = False)
planes_.head(1)

tailnum  created   age
119788  N381AA   1956.0  57.0

The oldest plane by tailnum is N381AA, which was created in 1956 and was 57 in 2013.

#### b. How many airplanes that flew from New York City are included in the planes table?

In [106]:
flights_ = pd.merge(flights, planes, left_on = 'tailnum', right_on = 'tailnum', how = 'inner')
flights_ = flights_.groupby('tailnum').sum().reset_index()
len(flights_)

3322

There are 3322 airplanes from the flights table that are included in the planes table.

### Question 3: Generate the code to convert the following data frame to wide format.

In [54]:
dat = pd.DataFrame({
    'grp':['A' , 'A' , 'B', 'B'],
    'sex':['F', 'M', 'F', 'M'],
    'meanL':[0.225, 0.47, 0.325, 0.547],
    'sdL':[0.106 ,.325, .106, .308],
    'meanR':[.34, .57, .4, .647],
    'sdR':[0.0849, 0.325, 0.0707, 0.274]
})
dat

grp sex  meanL    sdL  meanR     sdR
0   A   F  0.225  0.106  0.340  0.0849
1   A   M  0.470  0.325  0.570  0.3250
2   B   F  0.325  0.106  0.400  0.0707
3   B   M  0.547  0.308  0.647  0.2740

In [56]:
pd.pivot(dat, index = 'grp', columns = 'sex', values = ['meanL', 'sdL', 'meanR', 'sdR'])

meanL           sdL        meanR            sdR       
sex      F      M      F      M     F      M       F      M
grp                                                        
A    0.225  0.470  0.106  0.325  0.34  0.570  0.0849  0.325
B    0.325  0.547  0.106  0.308  0.40  0.647  0.0707  0.274

### Question 4: Consider the pccc icd10 dataset.

In [57]:
pccc_icd10_dataset = pd.read_csv("https://raw.githubusercontent.com/gjm112/DSCI401/main/data/pccc_icd10_dataset.csv")
pccc_icd10_dataset.head()

id      dx1      dx2      dx3      dx4      dx5      dx6      dx7      dx8  \
0   1  S9410XS   I67841   E70339      NaN  S14121A   M66229  S92065G    O0973   
1   2      NaN  S53422D  S92244B   M66342      NaN  S32442A  T1582XD  S72325C   
2   3      NaN  S91225S      NaN  W6119XD    C8397  M80819K  S72114R      NaN   
3   4  S7226XK    Y93G2    L0592   K08530      NaN  S62637D  T84612A      NaN   
4   5  S92246A    O4212    D2920  S42434S   F15980      NaN  S52572R  M8080XA   

       dx9  ...        g1        g2        g3        g4        g5        g6  \
0      NaN  ...  239196.0  672832.0  683784.0  757546.0       NaN  168052.0   
1  S52131B  ...  931331.0  404900.0  912213.0       NaN  964580.0  371556.0   
2  Y382X3D  ...  627455.0  638100.0  745829.0  843799.0  322975.0       NaN   
3      NaN  ...  809782.0  153243.0  413723.0  130995.0  211708.0  610135.0   
4  X731XXD  ...       NaN  636794.0       NaN  928572.0  930823.0  168586.0   

         g7        g8        g9       g10  
0  104625.0       NaN  850974.0       NaN  
1  778488.0  115827.0  440619.0  955264.0  
2       NaN  932106.0  289004.0  242699.0  
3       NaN  471383.0  191245.0  135116.0  
4  133292.0  699936.0  500743.0       NaN  

[5 rows x 31 columns]

#### a. Remove all the columns labeled with ”g” and a number.

In [59]:
pccc_icd10_dataset = pccc_icd10_dataset.loc[:, ~pccc_icd10_dataset.columns.str.startswith('g')]
pccc_icd10_dataset

id      dx1      dx2      dx3      dx4      dx5      dx6      dx7  \
0       1  S9410XS   I67841   E70339      NaN  S14121A   M66229  S92065G   
1       2      NaN  S53422D  S92244B   M66342      NaN  S32442A  T1582XD   
2       3      NaN  S91225S      NaN  W6119XD    C8397  M80819K  S72114R   
3       4  S7226XK    Y93G2    L0592   K08530      NaN  S62637D  T84612A   
4       5  S92246A    O4212    D2920  S42434S   F15980      NaN  S52572R   
..    ...      ...      ...      ...      ...      ...      ...      ...   
995   996      NaN  S12041S     K121  S0522XA  O4102X2      NaN  S72392A   
996   997  S43203S  O3130X2      NaN    M9959      NaN   M67372    Q7271   
997   998      NaN  S45802S  X718XXS  S6440XA   F19129   F14281  S0910XA   
998   999    C3481  S21252S      NaN   I69342  S15011D  S23420S  T8459XD   
999  1000     O070    Q6471      NaN      NaN  S61312S   Z13850  S82013F   

         dx8      dx9  ...      pc1      pc2      pc3      pc4      pc5  \
0      O0973      NaN  ...  0PSH3CZ  0JPT3XZ  037906Z  0JHD3HZ  0KQ54ZZ   
1    S72325C  S52131B  ...  0DVM7DZ  0NRJ47Z  DWY48ZZ  0HRWX7Z  BP091ZZ   
2        NaN  Y382X3D  ...  0PBV4ZX  0XM20ZZ  0DWD4UZ  2W07XYZ  F0636ZZ   
3        NaN      NaN  ...  DDY37ZZ  07LL0CZ  0Y9930Z  037M3GZ  04100Z4   
4    M8080XA  X731XXD  ...  02UL4KZ  03VD0ZZ  02110K8  3E050HZ  3E0U0GB   
..       ...      ...  ...      ...      ...      ...      ...      ...   
995   I87012  T484X3S  ...  0HN0XZZ  047R3D6  005P4ZZ  0PW1XKZ  041H4K9   
996     Z809      NaN  ...      NaN  0DQ70ZZ  06UN0JZ  03Q73ZZ  07L74ZZ   
997    M7661   M67823  ...  05BB0ZZ  0NSB0ZZ  0D957ZX  0CWY7KZ  07UH47Z   
998  V752XXS  S06383S  ...      NaN  0R954ZZ      NaN  0RU537Z  0PU64JZ   
999  T59894A  S52033P  ...  06923ZZ  0X340ZZ  09503ZZ      NaN  05R50JZ   

         pc6      pc7      pc8      pc9     pc10  
0    0WPK3YZ  01B04ZX  0DWV07Z  09513ZZ  0V554ZZ  
1    0Y0H4JZ      NaN  0B9880Z      NaN      NaN  
2    0RUP37Z      NaN  0WCP8ZZ  0DUM4KZ  BN02ZZZ  
3        NaN  0SPG33Z  0TRC07Z  041M0KQ  DB10B8Z  
4        NaN  0SPQ30Z  0WWBXYZ      NaN  0SWN38Z  
..       ...      ...      ...      ...      ...  
995  4A0D73Z  0QUM07Z  07L43ZZ      NaN  BR49ZZZ  
996  0VC53ZZ  2W08X5Z  0D5V0ZZ  0B9H8ZZ  0T184KD  
997  00XL4ZG  0GB10ZX  0S963ZZ  F0BZ0KZ      NaN  
998  03U337Z  03BQ0ZX  02VR3CT      NaN  0N940ZZ  
999  0BH082Z  01913ZX  05SY4ZZ  F07C6GZ  0NS5XZZ  

[1000 rows x 21 columns]

#### b. Convert this to a long data set with three columns: id, type (pc or dx), and code.

In [91]:
columns = pccc_icd10_dataset.columns.tolist()
del columns[0] #Removed 'id'
pd.melt(pccc_icd10_dataset, id_vars = 'id', value_vars = columns, var_name = "type", value_name = "code")

id  type     code
0         1   dx1  S9410XS
1         2   dx1      NaN
2         3   dx1      NaN
3         4   dx1  S7226XK
4         5   dx1  S92246A
...     ...   ...      ...
19995   996  pc10  BR49ZZZ
19996   997  pc10  0T184KD
19997   998  pc10      NaN
19998   999  pc10  0N940ZZ
19999  1000  pc10  0NS5XZZ

[20000 rows x 3 columns]